<h1>Python Queue</h1>

<b><i>A queue is defined as a linear data structure that is open at both ends and the operations are performed in First In First Out (FIFO) order.</i></b>

<i>We define a queue to be a list in which all additions to the list are made at one end, and all deletions from the list are made at the other end.  The element which is first pushed into the order, the operation is first performed on that.</i>

<i>A real life example of a queue can be seen in a ticket counter, where the person who gets in the line earlier, also leaves the line earlier when compared to people who got in the line later.</i>

<b><i>How do portals like Yahoo Finance or Money Control consume the price of stocks and make it available for us to track? These portals get their input feed from a stock exchange like Bombay Stock Exchange. Coming to the technical architecture, the engineering team at BSE will be continuously providing the prices of respective stocks to these portals. But how does a portal like Yahoo Finance consume this information being sent by BSE?</i></b>

<h5>Micro-service Architecture</h5>

<i>One way to receive the prices of respective stocks if for Yahoo Finance to build a HTTP server and BSE to continuouly make HTTP POST API synchronous calls to the HTTP server to communicate the information in the form of a JSON objects.</i>
    
<i>But what would happen if the HTTP server needs maintenance or is down for any reason for as little as 5 minutes? In that case, BSE would be making the HTTP POST API calls but it would only be resulting in loss of data - the prices that were posted during the time the HTTP server was down would be lost. For the portal, once the HTTP server comes back up after 5 minutes, there would be no way to retrieve the old lost prices for which BSE had made synchronous HTTP POST API calls.</i>

<i>Also, what if tomorrow Google Finance and Money Control also want to consume the prices of respective stocks from BSE?Now, the team at BSE have to update their code and add Google's and Money Control's URL to post the prices to their portals. This is called a tightly coupled architecture which has a lot of issues, like - every time a new consumer is to be on-boarded, the respective URL needs to be added by the producer. Also, if there is any change in the URL at the consumer end, the same has to be updated by the producer to ensure data is sent to the correct end-points.</i>

<h5>Queue Architecture</h5>

<i>What if there was a memory buffer where BSE could just upload the prices of respective stocks every minute for the consumers to consume? Portals like Yahoo Finance could just read the prices from the buffer in the order they were entered. Also, once Yahoo Finance consumes the price of a stock at a paticular time, their pointer will move to the next element to read the prices of the stocks at the next minute, leaving Google Finance and Money Control to consume prices that Yahoo Finance might already have consumed if their respective pointers point to those elements. This is possible because each portal will have their respective pointers which will consume the price of the respective stock at the minute the pointer is pointing.</i>

<b><i>This memory buffer we used and the architure it followed for inserting and consuming information is the implementation of a Stack data-structure.</i></b>

<i>It also enables loose coupling between the systems, i.e. tomorrow, if IND Money intends to get the prices of respective stocks from BSE, IND Money can start reading the information pushed by BSE in the memory buffer without the team at BSE requiring to perform any additional action. <b>A queue is thus a very flexible data-structure with very minimal issues.</b></i>

<i>The architecture we just talked about above is also called the producer-consumer problem, where one entity produce some information and some other entities are consuming the information produced by a producer in a way that they are not tightly coupled.</i>

<b><i>Here, whatever is pushed first in the buffer is consumed first, hence it is called FIFO (First In First Out) data-structure, unlike Stack.</i></b>

<img align="left" src="Queue Producer Consumer.png" alt="Stack LIFO" style="border: 5px solid #555;">

<h2>Python Queue ADT Implementation using Python List</h2>

In [1]:
stock_price_queue = []

In [2]:
#We always insert at the 0th index

stock_price_queue.insert(0, 131.10)
stock_price_queue.insert(0, 132.12)
stock_price_queue.insert(0, 135)

In [3]:
stock_price_queue

[135, 132.12, 131.1]

<i>We can see that 131.1 was the first element we inserted. However, as newer elements were pushed into the list after 131.1, 131.1 kept moving further down the list.<b> As we know that queue is a first-in first-out data-structure, 131.1 should be the first element we consume from this queue.</b></i>

In [4]:
stock_price_queue.pop()

131.1

In [5]:
stock_price_queue

[135, 132.12]

In [6]:
stock_price_queue.pop()

132.12

In [7]:
stock_price_queue

[135]

In [8]:
stock_price_queue.pop()

135

In [9]:
stock_price_queue

[]

In [10]:
stock_price_queue.pop() #Error/Exception - Nothing to pop from empty list

IndexError: pop from empty list

<i>As was the case with Stack ADT, Python Queue ADT can be implemented using Python lists, but are not the recommended approach for the implementation since Python lists have the constraint of being dynamic arrays, in which elements need to be copied from one memory area to a new memory area if the initial memory allocated for the list is exhausted. This can increase the time-complexity of operations to be performed on the queue.<b> The preffered/recommended implementation of Queues in Python is using the deque module from Collections class.</b></i>

<h2>Python Queue ADT Implementation using collections.deque()</h2>

<b><i>collections.deque() is Python's specific generalisation of Stacks and Queues which is itself implemented using Doubly Linked Lists.</i></b>

<i>Reference to Python's <a href="https://docs.python.org/3/library/collections.html#collections.deque" target="_blank">collections.deque()</a> from Python Docs.</i>

In [11]:
from collections import deque

q = deque()

In [12]:
#For Stack ADT, we always append at the end, but for Queue ADT, we append at the beginning and remove from the end

#Check available functions it supports for appending to the beginning
dir(q)

['__add__',
 '__bool__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__copy__',
 '__delattr__',
 '__delitem__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__gt__',
 '__hash__',
 '__iadd__',
 '__imul__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__reversed__',
 '__rmul__',
 '__setattr__',
 '__setitem__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'append',
 'appendleft',
 'clear',
 'copy',
 'count',
 'extend',
 'extendleft',
 'index',
 'insert',
 'maxlen',
 'pop',
 'popleft',
 'remove',
 'reverse',
 'rotate']

<i>Referencing the available functions with the documentation of collections.deque(), we would be needing the <b>appendleft()</b> function to push to the queue and <b>pop()</b> to remove from the same.</i>

In [13]:
q.appendleft(5)
q.appendleft(10)
q.appendleft(15)

In [14]:
q

deque([15, 10, 5])

In [15]:
q.pop() #Should pop out 5 as queue is FIFO

5

In [16]:
q

deque([15, 10])

In [17]:
q.pop() #Should return 10

10

In [18]:
q

deque([15])

In [19]:
q.pop()

15

In [20]:
q

deque([])

In [21]:
q.pop() #Error/Exception - Can't pop from empty queue

IndexError: pop from an empty deque

<h2>Implement Stack ADT using collections.deque()</h2>

In [22]:
#Custom implementation to follow the proper Stack template prototype - enqueue(value), dequeue(), is_empty(), size()

from collections import deque

class Queue:
    def __init__(self):
        self.buffer = deque()
        
    def enqueue(self, value):
        self.buffer.appendleft(value)
        
    def dequeue(self):
        return self.buffer.pop()
        
    def is_empty(self):
        return (len(self.buffer) == 0)
    
    def size(self):
        return len(self.buffer)

In [23]:
q = Queue()

In [24]:
q.buffer #Empty Queue

deque([])

In [25]:
q.is_empty()

True

In [26]:
q.size()

0

In [27]:
#Enqueue into queue - queue of dictionaries

q.enqueue({
    "company": "Reliance Inc.",
    "timestamp": "2021/12/01 11:00:00",
    "price": 131.10
})

In [28]:
q.buffer

deque([{'company': 'Reliance Inc.',
        'timestamp': '2021/12/01 11:00:00',
        'price': 131.1}])

In [29]:
q.enqueue({
    "company": "Reliance Inc.",
    "timestamp": "2021/12/01 11:01:00",
    "price": 132.12
})

q.enqueue({
    "company": "Reliance Inc.",
    "timestamp": "2021/12/01 11:03:00",
    "price": 135
})

In [30]:
q.buffer

deque([{'company': 'Reliance Inc.',
        'timestamp': '2021/12/01 11:03:00',
        'price': 135},
       {'company': 'Reliance Inc.',
        'timestamp': '2021/12/01 11:01:00',
        'price': 132.12},
       {'company': 'Reliance Inc.',
        'timestamp': '2021/12/01 11:00:00',
        'price': 131.1}])

<i>We inserted price with timestamp of 11:00:00 before the other 2 prices, so it is at the end of the queue where the operation will be performed earlier.</i>

In [31]:
q.dequeue()

{'company': 'Reliance Inc.',
 'timestamp': '2021/12/01 11:00:00',
 'price': 131.1}

In [32]:
q.buffer

deque([{'company': 'Reliance Inc.',
        'timestamp': '2021/12/01 11:03:00',
        'price': 135},
       {'company': 'Reliance Inc.',
        'timestamp': '2021/12/01 11:01:00',
        'price': 132.12}])

In [33]:
q.size()

2

<i>In out example above, the various platforms like Yahoo Finance and Google Finance will perform the dequeue operation for their respective pointer and the team at BSE will perform the enqueue operation.</i>

In [34]:
q.dequeue()

{'company': 'Reliance Inc.',
 'timestamp': '2021/12/01 11:01:00',
 'price': 132.12}

In [35]:
q.dequeue()

{'company': 'Reliance Inc.', 'timestamp': '2021/12/01 11:03:00', 'price': 135}

In [36]:
q.buffer

deque([])

In [37]:
q.dequeue() #Error - Empty queue

IndexError: pop from an empty deque

In [38]:
q.size()

0

In [39]:
q.is_empty()

True

<b><i>The worst time-complexity of the legal insertion and deletion operations in a queue is O(1)</i></b>

<i>Search and access operations are not allowed operations in a Queue ADT, however they can be performed at O(n) worst-case time-complexity.</i>

<h2>Exercise 1</h2>

Using the Queue class from above, design a food ordering system where your Python program will run two threads:

1. Place Order: This thread will be placing an order and inserting that into a queue. This thread places new order every 0.5 second. (hint: use time.sleep(0.5) function) <br>
2. Serve Order: This thread will server the order. All you need to do is pop the order out of the queue and print it. This thread serves an order every 2 seconds. Also start this thread 1 second after place order thread is started.

Use this video to get yourself familiar with <a href="https://www.youtube.com/watch?v=PJ4t2U15ACo&list=PLeo1K3hjS3uub3PRhdoCTY8BxMKSW7RjN&index=3" target="_blank">Multithreading in Python</a>.

Pass following list as an argument to place order thread: <br>
orders = ['pizza','samosa','pasta','biryani','burger']

This problem is a producer, consumer problem where place_order thread is producing orders whereas server_order thread is consuming the food orders.

In [40]:
from collections import deque

class Queue:
    def __init__(self):
        self.buffer = deque()
        
    def enqueue(self, value):
        self.buffer.appendleft(value)
        
    def dequeue(self):
        return self.buffer.pop()
        
    def is_empty(self):
        return (len(self.buffer) == 0)
    
    def size(self):
        return len(self.buffer)

In [41]:
#Without multi-threading

orders_queue = Queue()

import time

def place_order(orders):
    for order in orders:
        print("Placing order for:", order)
        orders_queue.enqueue(order)
        time.sleep(0.5)
        
def serve_order():
    time.sleep(1)
    while not orders_queue.is_empty():
        print("Serving order for:", orders_queue.dequeue())
        time.sleep(2)

orders = ['pizza','samosa','pasta','biryani','burger']

t = time.time()

place_order(orders)
serve_order()

print("Completed in:", time.time() - t, "seconds")

Placing order for: pizza
Placing order for: samosa
Placing order for: pasta
Placing order for: biryani
Placing order for: burger
Serving order for: pizza
Serving order for: samosa
Serving order for: pasta
Serving order for: biryani
Serving order for: burger
Completed in: 13.539191007614136 seconds


In [42]:
#With multi-threading

orders_queue = Queue()

import time
import threading

def place_order(orders):
    for order in orders:
        print("Placing order for:", order)
        orders_queue.enqueue(order)
        time.sleep(0.5)
        
def serve_order():
    time.sleep(1)
    while not orders_queue.is_empty():
        print("Serving order for:", orders_queue.dequeue())
        time.sleep(2)

orders = ['pizza','samosa','pasta','biryani','burger']

t = time.time()

thread1 = threading.Thread(target=place_order, args=(orders,))
thread2 = threading.Thread(target=serve_order)

thread1.start()
thread2.start()

thread1.join()
thread2.join()

print("Completed in:", time.time() - t, "seconds")

Placing order for: pizza
Placing order for: samosa
Serving order for: pizza
Placing order for: pasta
Placing order for: biryani
Placing order for: burger
Serving order for: samosa
Serving order for: pasta
Serving order for: biryani
Serving order for: burger
Completed in: 11.022597074508667 seconds


<h2>Exercise 2</h2>

Write a program to print binary numbers from 1 to 10 using Queue. Use Queue class implemented above. Binary sequence should look like: <br>
    1 <br>
    10 <br>
    11 <br>
    100 <br>
    101 <br>
    110 <br>
    111 <br>
    1000 <br>
    1001 <br>
    1010
    
<u>Hint:</u> Notice a pattern above. After 1, second and third number is 1+0 and 1+1. 4th and 5th number are second number (i.e. 10) + 0 and second number (i.e. 10) + 1.

Also add front() function in queue class that can return the front element in the queue.

In [43]:
from collections import deque

class Queue:
    def __init__(self):
        self.buffer = deque()
        
    def enqueue(self, value):
        self.buffer.appendleft(value)
        
    def dequeue(self):
        if not self.is_empty():
            return self.buffer.pop()
        
    def is_empty(self):
        return (len(self.buffer) == 0)
    
    def size(self):
        return len(self.buffer)
    
    def front(self):
        return self.buffer[-1]

In [44]:
def produce_binary_numbers(n):
    number_queue = Queue()
    number_queue.enqueue("1")
    
    for i in range(n):
        front = number_queue.front()
        print(front)
        
        number_queue.enqueue(str(front) + "0")
        number_queue.enqueue(str(front) + "1")
        
        number_queue.dequeue()
        
produce_binary_numbers(10)

1
10
11
100
101
110
111
1000
1001
1010
